docs/with-js.md

```diff
@@ -0,0 +1,173 @@
+  ([a, b]) => a || b
+);
+
+/**
+ * @typedef SimpleComponentSignature
+ * @property {object} Args
+ * @property {string} Args.message
+ */
+
+/** @extends {Component<SimpleComponentSignature>} */
+export default class SimpleComponent extends Component {
+  static template = hbs`
+    <h1>This is my simple message: {{or @message 'hello'}}</h1>
+  `;
+}
+```
+
+### Signatures
+
+It is possible write a full `Signature` type for a JS-only component. This will provide a useful hook not only for Glint, but also for documentation tools which can consume JSDoc. Here's an example of a fully-specified `Signature` for a JS component.
+
+```js
+import Component from '@glimmer/component';
+
+/**
+ * @typedef FancyInputArgs
+ * @property {'input' | 'textarea'} type The type of input to render
+ * @property {string} value The initial value to render into the input
+ * @property {(newValue: string) => void} onChange An action to run when the
+ *   input's value changes
+ */
+
+/**
+ * @typedef {HTMLInputElement | HTMLTextAreaElement} FancyInputElement
+ */
+
+/**
+ * @typedef FancyInputBlocks
+ * @property {[]} label
+ */
+
+/**
+ * @typedef FancyInputSignature
+ * @property {FancyInputArgs} Args
+ * @property {FancyInputElement} Element
+ * @property {FancyInputBlocks} Blocks
+ */
+
+/**
+ * A fancy `<input>` component, with styles pre-applied and some custom
+ * handling.
+ *
+ * @extends {Component<FancyInputSignature>}
+ */
+export default class FancyInput extends Component {
+  // implementation...
+}
+```
+
+Let's walk through this in detail. First, the `Args` to the component specify that it receives:
+
+- a `@type` argument specifying whether to render an `<input type='text'>` or a `<textarea>`
+- a `@value` argument, which should be a string
+- an `@onChange` action which receives the updated value and does something with it
+
+```js
+/**
+ * @typedef FancyInputArgs
+ * @property {'input' | 'textarea'} type The type of input to render
+ * @property {string} value The initial value to render into the input
+ * @property {(newValue: string) => void} onChange An action to run when the
+ *   input's value changes
+ */
+```
+
+The `@typedef` declaration for an `Element` must be a "union" type, declaring all the different types of element to which `...attributes` might be applied, with a pipe `|` as the "or" character for the different types. Here, we can imagine that the `...attributes` are applied to the input, and since it might be an `<input>` or a `<textarea>`, we write exactly that, using the names for the corresponding JS types:
+
+```js
+/**
+ * @typedef {HTMLInputElement | HTMLTextAreaElement} FancyInputElement
+ */
+```
+
+The component also provides a named block, `label`, which does not yield any value, allowing customization of the `<label>` for the `<input>`:
+
+```js
+/**
+ * @typedef FancyInputBlocks
+ * @property {[]} label
+ */
+```
+
+Finally, the component signature itself assembles those and also specifies the `Element`:
```

Comment:
Are we missing an actual `@typedef` for `FancyComponentSignature` or am I just glossing over it?

Reply:
Nope, you’re correct: I forgot to copy it down here!